File: go/tests/pressure/lps28dfw_test/main.go
//go:build linux && !tinygo

// LPS28DFW hardware test — Linux host.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/pressure"
	"github.com/tuhde/Periph/go/periph/connection"
)

func envOr(k, d string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return d
}

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x5C"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	chip, err := pressure.NewLPS28DFWMinimal(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "init:", err)
		os.Exit(2)
	}

	t, err := chip.ReadTemperature()
	check("temperature_range", err == nil && t >= -40.0 && t <= 85.0)
	p, err := chip.ReadPressure()
	check("pressure_range", err == nil && p >= 260.0 && p <= 1260.0)

	chipFull, err := pressure.NewLPS28DFWFull(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "init full:", err)
		os.Exit(2)
	}
	if err := chipFull.Configure(pressure.LPS28DFWODR100Hz, pressure.LPS28DFWAVG128,
		pressure.LPS28DFWFSMode2, pressure.LPS28DFWLFPFODROver9, false); err == nil {
		check("full_configure", true)
	} else {
		check("full_configure", false)
	}

	alt, err := chipFull.Altitude(1013.25)
	check("altitude_range", err == nil && alt >= -500.0 && alt <= 9000.0)

	cid, err := chipFull.ChipID()
	check("chip_id_or_no_chip", err == nil && (cid == 0xB4 || cid == 0x00))

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

File: rust/tests/pressure/lps28dfw_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::pressure::{Lps28dfwMinimal, Lps28dfwFull, LPS28DFW_ODR_100_HZ, AVG_128, FS_MODE_2};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x5C);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut lps = Lps28dfwMinimal::new(dev, addr).expect("init LPS28DFW");

    let t = lps.read_temperature().unwrap_or(0.0);
    check_true!(t >= -40.0 && t <= 85.0, "temperature_range", passed, failed);

    let p = lps.read_pressure().unwrap_or(0.0);
    check_true!(p >= 260.0 && p <= 1260.0, "pressure_range", passed, failed);

    drop(lps);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut lps_full = Lps28dfwFull::new(dev, addr).expect("init LPS28DFW Full");

    lps_full.configure(LPS28DFW_ODR_100_HZ, AVG_128, FS_MODE_2, false, 1).unwrap();
    check_true!(lps_full.read_pressure().is_ok(), "configure", passed, failed);

    let threshold_raw = (1050.0_f32 * 16.0_f32) as i32;
    let capped = if threshold_raw > 0x7FFF { 0x7FFF } else { threshold_raw };
    check_true!(capped == 16800, "threshold_raw_conversion", passed, failed);

    let alt = lps_full.altitude(1013.25).unwrap_or(-1.0);
    check_true!(alt >= -500.0 && alt <= 9000.0, "altitude_range", passed, failed);

    let cid = lps_full.chip_id().unwrap_or(0);
    check_true!(cid == 0xB4 || cid == 0x00, "chip_id_or_no_chip", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}